File: sdks/python/examples/watch_run.py
"""Drive one run from the Python SDK: create it, watch it, answer it.

    node sdks/examples/runs-wire-stub.mjs &
    python sdks/python/examples/watch_run.py

Point it at a studio daemon instead with WIGOLO_BASE_URL / WIGOLO_API_TOKEN; nothing in this
file changes.
"""

from __future__ import annotations

import os
import sys
import threading
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "src"))

from wigolo import Client  # noqa: E402

client = Client(base_url=os.environ.get("WIGOLO_BASE_URL", "http://127.0.0.1:8787"))

run = client.runs.create(
    task="compare three 27-inch monitors",
    driver={"kind": "sdk", "client": {"name": "demo", "version": "1.0"}},
)
print(f"run {run['id']} — {run['status']} — driver {run['driver']['kind']}")

finished = threading.Event()


def on_human_message(event):
    print(f"  [human]    seq {event['seq']}: {event['payload']['text']}")


def on_approval(event):
    anchor = event["payload"].get("anchor", {})
    print(
        f"  [approval] seq {event['seq']}: {event['payload']['prompt']} "
        f"(anchored to mark {anchor.get('mark')} on {anchor.get('tabId')})"
    )


def on_takeover(event):
    print(f"  [takeover] seq {event['seq']}: {event['payload']['reason']}")


def on_event(event):
    print(f"  [event]    seq {event['seq']}: {event['type']}")
    if event["type"] == "run.completed":
        finished.set()


watch = client.runs.watch_run(
    run["id"],
    on_human_message=on_human_message,
    on_approval=on_approval,
    on_takeover=on_takeover,
    on_event=on_event,
    on_error=lambda exc, event: print(f"  [error]    {exc}"),
    reconnect_delay=0.1,
)

# A message is QUEUED, never sent. The server's own state line says so; print it verbatim.
queued = client.runs.send_message(run["id"], text="prefer 120Hz panels")
print(f"message {queued['message']['message_id']}: {queued['message']['state_line']}")

# Request-the-wheel is a gesture, never a race.
gesture = client.runs.driver_gesture(
    run["id"],
    gesture="request",
    by={"kind": "sdk", "client": {"name": "demo", "version": "1.0"}},
    reason="I can finish the comparison",
)
print(
    f"wheel requested — requestId {gesture.get('requestId')}, "
    f"events {len(gesture['events'])}"
)

finished.wait(timeout=30)
watch.stop()
watch.join(timeout=10)
print(f"watch stopped at seq {watch.last_seq}")

File: sdks/typescript/examples/watch-run.mjs
/**
 * Drive one run from the TypeScript SDK: create it, watch it, answer it.
 *
 *   node sdks/examples/runs-wire-stub.mjs &
 *   node sdks/typescript/examples/watch-run.mjs
 *
 * Point it at a studio daemon instead with WIGOLO_BASE_URL / WIGOLO_API_TOKEN; nothing in this
 * file changes. Run it from the repo root after `npm run build --prefix sdks/typescript`.
 */
import { WigoloClient } from '../dist/index.js';

const client = new WigoloClient({
  baseUrl: process.env.WIGOLO_BASE_URL ?? 'http://127.0.0.1:8787',
});

const run = await client.runs.create({
  task: 'compare three 27-inch monitors',
  driver: { kind: 'sdk', client: { name: 'demo', version: '1.0' } },
});
console.log(`run ${run.id} — ${run.status} — driver ${run.driver.kind}`);

const watch = client.runs.watchRun(
  run.id,
  {
    onHumanMessage: (event) => {
      console.log(`  [human]    seq ${event.seq}: ${event.payload.text}`);
    },
    onApproval: async (event) => {
      const anchor = event.payload.anchor;
      console.log(
        `  [approval] seq ${event.seq}: ${event.payload.prompt} ` +
          `(anchored to mark ${anchor?.mark} on ${anchor?.tabId})`,
      );
    },
    onTakeover: (event) => {
      console.log(`  [takeover] seq ${event.seq}: ${event.payload.reason}`);
    },
    onEvent: (event) => {
      console.log(`  [event]    seq ${event.seq}: ${event.type}`);
      if (event.type === 'run.completed') watch.stop();
    },
    onError: (err) => console.log(`  [error]    ${err}`),
  },
  { reconnectDelayMs: 100 },
);

// A message is QUEUED, never sent. The server's own state line says so and we print it verbatim.
const queued = await client.runs.sendMessage(run.id, { text: 'prefer 120Hz panels' });
console.log(`message ${queued.message.message_id}: ${queued.message.state_line}`);

// Request-the-wheel is a gesture, never a race.
const gesture = await client.runs.driverGesture(run.id, {
  gesture: 'request',
  by: { kind: 'sdk', client: { name: 'demo', version: '1.0' } },
  reason: 'I can finish the comparison',
});
console.log(`wheel requested — requestId ${gesture.requestId}, events ${gesture.events.length}`);

await watch.done;
console.log(`watch stopped at seq ${watch.lastSeq}`);
